File: haskell/glue/src/Glue/Lib/List/Append.hs
module Glue.Lib.List.Append where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

append :: IR Eval
append = NativeFunc appendImpl

appendImpl :: IR Eval -> Eval (IR Eval)
appendImpl list = pure $ NativeFunc (appendTo list)

appendTo :: IR Eval -> IR Eval -> Eval (IR Eval)
appendTo list item = case list of
    List xs -> pure $ List (xs <> [item])
    _ -> throwError $ wrongArgumentType ["list", "item"]

File: dart/glue/lib/src/lib/list/append.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Prep function - preptructs a new list by appending an element to a list
/// Mirrors Haskell Glue.Lib.List.Append.append exactly
Ir append = IrNativeFunc(appendImpl);

/// Prep function implementation
/// Mirrors Haskell Glue.Lib.List.Append.appendImpl exactly
Eval<Ir> appendImpl(Ir list) {
  return Eval.pure(IrNativeFunc(appendWith(list)));
}

/// Helper function for tail argument
/// Mirrors Haskell Glue.Lib.List.Append.appendWith exactly
Eval<Ir> Function(Ir) appendWith(Ir list) {
  return (Ir item) => switch (list) {
    IrList(elements: final elements) => Eval.pure(IrList([...elements, item])),
    _ => throwError(wrongArgumentType(['list', 'any'])),
  };
}
